daemon/container_operations.go

```diff
@@ -1069,13 +1069,27 @@ func (daemon *Daemon) ConnectToNetwork(ctx context.Context, ctr *container.Conta
 		}
 
 		n, err := daemon.FindNetwork(idOrName)
-		if err == nil && n != nil {
+		if err == nil {
+			if _, ok := ctr.NetworkSettings.Networks[idOrName]; ok {
+				return fmt.Errorf("endpoint with name %s already exists in network %s", strings.TrimPrefix(ctr.Name, "/"), idOrName)
+			}
+
 			if err := daemon.updateNetworkConfig(ctr, n, endpointConfig, true); err != nil {
 				return err
 			}
 		} else {
-			ctr.NetworkSettings.Networks[idOrName] = &network.EndpointSettings{
-				EndpointSettings: endpointConfig,
+			_, errNr := daemon.cluster.GetNetwork(idOrName)
+
+			if errNr == nil {
+				if _, ok := ctr.NetworkSettings.Networks[idOrName]; ok {
```

Comment:
Because it's using `idOrName` this will miss a network that's added by name then id. For example...

```
# docker network create n2
d0d63bb13b7ce4e2b70daa2c97f35719b359898803e6a2e1db4aac66f592d594
# docker network connect n2 c1
# docker network connect n2 c1
Error response from daemon: endpoint with name c1 already exists in network n2
# docker network connect d0d63bb c1
#
```

If we're going this way, it'd probably be worth trying to catch that too - perhaps we should always store the connection under network name if the network already exists.

It's currently possible to run a container connected to a network by-name. Then, stop the container, delete and re-create the network with the same name. Then restart the container. That doesn't work if the container was originally created with its network specified by id rather than name. So, the id/name confusion isn't a new problem. But, insisting on networks existing before they can be used in container create/connect commands would make it fixable.

(We'd also want to rejig this a little, to [indent the error flow](https://go.dev/wiki/CodeReviewComments#indent-error-flow).)

Reply:
> That doesn't work if the container was originally created with its network specified by id rather than name

If we connect the network to the container, either by ID or by name, the container ultimately does not start:

```
❯ docker network create n1
55371e7e343ea0ef85f7eb2e108df24650066f2c96aded3416eed1b024e3e826
❯ docker network connect n1 c1
❯ docker stop c1
c1
❯ docker network rm n1
n1
❯ docker network create n1
9fab8485e6d4035b92c27e9573a2e2d898902fcbe811f3820465b35772af44e2
❯ docker start c1
Error response from daemon: Could not attach to network 55371e7e343ea0ef85f7eb2e108df24650066f2c96aded3416eed1b024e3e826: rpc error: code = NotFound desc = network 55371e7e343ea0ef85f7eb2e108df24650066f2c96aded3416eed1b024e3e826 not found
Error: failed to start containers: c1
❯ docker network create n2
58ea8a064a6d883c5d2b6511b7b790e0ff8c6b072d269c65a28258658099479a
❯ docker network connect 58ea8a c2
❯ docker stop c2
c2
❯ docker network rm n2
n2
❯ docker network create n2
c1b183594d87c5f1ce850dc8bee80895cdad8ed6212b3f6e0a6ef2897b70d0b8
❯ docker start c2
Error response from daemon: Could not attach to network 58ea8a064a6d883c5d2b6511b7b790e0ff8c6b072d269c65a28258658099479a: rpc error: code = NotFound desc = network 58ea8a064a6d883c5d2b6511b7b790e0ff8c6b072d269c65a28258658099479a not found
Error: failed to start containers: c2
```